[SPARK-15353] [CORE] Making peer selection for block replication pluggable

shubhamchopra opened the pull request:
## What changes were proposed in this pull request?

This PR makes block replication strategies pluggable. It provides two trait that can be implemented, one that maps a host to its topology and is used in the master, and the second that helps prioritize a list of peers for block replication and would run in the executors.

This patch contains default implementations of these traits that make sure current Spark behavior is unchanged.
## How was this patch tested?

This patch should not change Spark behavior in any way, and was tested with unit tests for storage.

shubhamchopra committed:
Adding capability to prioritize peer executors based on rack awareness while replicating blocks.

shubhamchopra committed:
Minor modifications to get past the style check errors.

shubhamchopra committed:
Using blockId hashcode as a source of randomness, so we don't keep choosing the same peers for replication.

shubhamchopra committed:
Several changes:
1. Adding rack attribute to hashcode and equals to block manager id.
2. Removing boolean check for rack awareness. Asking master for rack info, and master uses topology mapper.
3. Adding a topology mapper trait and a default implementation that block manager master endpoint uses to discern topology information.

shubhamchopra committed:
Adding null check so a Block Manager can be initiaziled without the master.

shubhamchopra committed:
Renaming classes/variables from rack to a more general topology.

shubhamchopra committed:
Renaming classes/variables from rack to a more general topology.

shubhamchopra committed:
We continue to randomly choose peers, so there is no change in current behavior.

shubhamchopra committed:
Spelling correction and minor changes in comments to use a more general topology instead of rack.

shubhamchopra committed:
Minor change. Changing replication info message to debug level.

shubhamchopra committed:
Providing peersReplicateTo to the prioritizer.

AmplabJenkins commented:
Can one of the admins verify this patch?

shubhamchopra committed:
Adding developer api annotations to TopologyMapper and BlockReplicationPrioritization

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala:
```diff
@@ -103,8 +121,11 @@ private[spark] object BlockManagerId {
+  def apply(execId: String,
+            host: String,
+            port: Int,
+            topologyInfo: Option[String] = None): BlockManagerId =
```
Maybe we need to correct indentation as below..

```scala
def apply(execId: String,
  host: String,
  port: Int,
  topologyInfo: Option[String] = None): BlockManagerId =
```

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -124,6 +138,13 @@ class BlockManagerMasterEndpoint(
       }
+
+    case GetTopologyInfo(host) => context.reply(getTopologyInfoForHost(host))
+
```
Maybe remove a newline here..

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,72 @@
+import org.apache.spark.annotation.DeveloperApi
+
+import scala.util.Random
+import org.apache.spark.internal.Logging
```
Maybe reorder imports.. (See [scala-style-guide#imports](https://github.com/databricks/scala-style-guide#imports))

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,72 @@
+   */
+  def prioritize(peers: Seq[BlockManagerId],
+                 peersReplicatedTo: Set[BlockManagerId],
+                 blockId: BlockId): Seq[BlockManagerId]
```
Maybe indentation as below:

```scala 
def prioritize(peers: Seq[BlockManagerId],
  peersReplicatedTo: Set[BlockManagerId],
  blockId: BlockId): Seq[BlockManagerId]
```

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,72 @@
+   */
+  override def prioritize(peers: Seq[BlockManagerId],
+                          peersReplicatedTo: Set[BlockManagerId],
+                          blockId: BlockId): Seq[BlockManagerId] = {
```
Here too..

```scala
override def prioritize(peers: Seq[BlockManagerId],
  peersReplicatedTo: Set[BlockManagerId],
  blockId: BlockId): Seq[BlockManagerId] = {
```

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
+    blockId: BlockId,
+    data: ChunkedByteBuffer,
+    level: StorageLevel,
+    classTag: ClassTag[_]): Unit = {
```
It seems this indentation might not have to be changed.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
+
+      if (numFailures > maxReplicationFailures
+                || peersForReplication.isEmpty
+                || peersReplicatedTo.size == numPeersToReplicateTo) {
```
Although I cannot find a documentation for indentation about this, it might be better be consistent because I see you added

```scala
executorId == id.executorId &&
  port == id.port &&
  host == id.host &&
  topologyInfo == id.topologyInfo
```

So this might have to be

```scala
if (numFailures > maxReplicationFailures
    || peersForReplication.isEmpty
    || peersReplicatedTo.size == numPeersToReplicateTo) {
```

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
+            peersReplicatedTo + peer,
+            peersFailedToReplicateTo)
+        } catch {
+          case e: Exception =>
```
I just wonder if it is safe to catch all the exceptions here..

shubhamchopra replied:
The intended behavior here is to keep trying with other available options, if there is any exception here, till we run out of options or have a high number of failures. The method does a "best-effort" replication in that sense. It was the same behavior in the earlier implementation as well.

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
+            // we have a failed replication, so we get the list of peers again
+            // we don't want peers we have already replicated to and the ones that
+            // have failed previously
+            val filteredPeers = getPeers(true).filter{p =>
```
Maybe as below,  (there is a good exmaple here, [scala-style-guide#pattern-matching](https://github.com/databricks/scala-style-guide#pattern-matching))
```scala
val filteredPeers = getPeers(true).filter { p =>
```

HyukjinKwon commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1079,109 +1103,97 @@ private[spark] class BlockManager(
+    val peersReplicatedTo = replicateBlock(0,
+      blockReplicationPrioritizer.prioritize(getPeers(false), Set.empty, blockId),
+      Set.empty,
+      Set.empty)
```
Maybe this should be as below:

```scala
replicateBlock(
  0,
  blockReplicationPrioritizer.prioritize(getPeers(false), Set.empty, blockId),
  Set.empty,
  Set.empty)
```

shubhamchopra committed:
Changes recommended by @HyukjinKwon to fix style issues.

shubhamchopra commented:
Fixed style issues pointed out by @HyukjinKwon

shubhamchopra committed:
Updating prioritizer api to use current blockmanager id for self identification.

ericl commented:
A couple high level questions:
- Rather than send an RPC to the master asking for a worker's topology info, is it possible for this to be provided at initialization time or determined based on the environment?

- Is it possible to narrow the interface of the prioritizer to just choose a single next peer? If it is desired to cache the prioritization order, this can be done internally within the prioritizer. For example, the interface could be something like this. Then the default prioritizer does not need to do a random shuffle of the entire peer list to choose its target.

```
trait BlockReplicationStrategy {

  trait ReplicationTargetSelector {
    def getNextPeer(
      candidatePeers: Set[BlockManagerId],
      successfulReplications: Set[BlockManagerId],
      failedReplications: Set[BlockManagerId]): Option[BlockManagerId]
  }

  def getTargetSelector(
    localId: BlockManagerId,
    blockId: BlockId,
    level: StorageLevel): ReplicationTargetSelector
}
```

Also, the patch would be more minimal if only the `getRandomPeer()` call was changed.

shubhamchopra commented:
The topology info is only queried when the executor initiates and is assumed to stay the same throughout the life of the executor. Depending on the cluster manager being used, I am assuming the exact way this information is provided may differ. Resolving this at the master makes this implementation simpler as only the master needs to be able to access the service/script/class being used to resolve the topology. The communication overhead is minimal as the executors do have to communicate with the master when they initiate anyways.

The getRandomPeer() method was doing quite a bit more than just getting a random peer. It was being used to manage/mutate state, which was being mutated in other places as well. I tried to keep the block placement strategy and the usage of its output separate, to make it simpler to provide a new block placement strategy. I also thought it would be best to de-couple any internal replication state management with the block replication strategy, while still keeping the structure of the state the same.

The costlier operation here is the RPC fetch of all the peers from the master. The prioritization algorithm is only called once if there are no failures.  If there are failures, the list of peers is requested from the master again, before the prioritizer is run. The bigger hit again, would be the RPC communication between the executor and the master. Random.shuffle in the default prioritizer uses Fisher-Yates shuffle, so is linear in time.

ericl commented:
> The topology info is only queried when the executor initiates and is assumed to stay the same throughout the life of the executor. Depending on the cluster manager being used, I am assuming the exact way this information is provided may differ. Resolving this at the master makes this implementation simpler as only the master needs to be able to access the service/script/class being used to resolve the topology. The communication overhead is minimal as the executors do have to communicate with the master when they initiate anyways.

I see, that makes sense, though it is a little weird to ask the master for info that you use to register right away later.

> The getRandomPeer() method was doing quite a bit more than just getting a random peer. It was being used to manage/mutate state, which was being mutated in other places as well. I tried to keep the block placement strategy and the usage of its output separate, to make it simpler to provide a new block placement strategy. I also thought it would be best to de-couple any internal replication state management with the block replication strategy, while still keeping the structure of the state the same.

Still, I think it would be a smaller change to just move some of that logic out of getRandomPeer(), and retain the rest. Then you just need to implement getNextPeer(), and BlockManager doesn't need to worry about tracking the prioritized order internally.

shubhamchopra commented:
The state being managed inside getRandomPeer() is also modified in a couple of other places, so it won't be a very clean change to remove some of it out of getRandomPeer. Even if that is done, I agree that your approach would only mean calling getNextPeer. It would however mean adding adding more state to ensure expected behavior in cases where block replication fails on a peer. 

I am flexible about the implementation choices, so can do the modifications if needed. Just to clarify on the motivation of this interface, I have another PR  [SPARK-15354](https://github.com/apache/spark/pull/13932) that shows a couple of prioritizers that I intend to add (including a simple one that replicates HDFS's block replication strategy). Note that in case of failures, the list of peers is requested from the master afresh and is optimized over again and with this interface. Let me know what you think.

ericl commented:
You wouldn't have to create a new selector after a failure. That case can be detected by checking if the number of failed replications has increased, e.g. `if (failedReplications.length > prevNumFails) { reprioritize... }`. Basically, that state would be tracked in the selector, instead of BlockManager.

I think the main benefit here is that the interface is more flexible as a developer facing API.

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -159,8 +162,27 @@ private[spark] class BlockManager(
+
     blockManagerId = BlockManagerId(
-      executorId, blockTransferService.hostName, blockTransferService.port)
+      executorId, blockTransferService.hostName, blockTransferService.port, topologyInfo)
```
Would it work if topologyInfo was sent back from the master when `registerBlockManager` is called? It doesn't seem that anything uses `blockManagerId` until registration finishes. That way we wouldn't need this weird two-step registration.

shubhamchopra replied:
I implemented this in the commits below.

shubhamchopra committed:
BlockManagerInitialization now only uses a single message to master to get a fully fleshed out id, with topology information, if available.

shubhamchopra committed:
1. Changing tail recursive function in BlockManager to an imperative while loop, as suggested by @ericl
2. Adding SparkConf constructor arguments to TopologyMapper, so any required properties like classname or file/script names can be passed to a custom topology mapper.

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,88 @@ private[spark] class BlockManager(
+          // we don't want peers we have already replicated to and the ones that
+          // have failed previously
+          val filteredPeers = getPeers(true).filter { p =>
+            !(peersFailedToReplicateTo.contains(p) || peersReplicatedTo.contains(p))
```
nit: seems a little easier to read if it was !a && !b

shubhamchopra replied:
Fixed this.

ericl replied:
?

ericl replied:
Still seems the same

shubhamchopra replied:
My bad, the latest commit fixed this. The while loop also had a similar condition and had fixed that earlier.

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala:
```diff
@@ -69,24 +72,39 @@ class BlockManagerId private (
     out.writeInt(port_)
+    out.writeBoolean(topologyInfo_.isDefined)
+    // if we don't keep topology information, we just write an empty string.
+    out.writeUTF(topologyInfo_.getOrElse(""))
```
You can just omit this write if it's missing?

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -50,12 +50,20 @@ class BlockManagerMaster(
   }
 
-  /** Register the BlockManager's id with the driver. */
+  /** Register the BlockManager's id with the driver. The input BlockManagerId does not contain
```
use javadoc style here

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -50,12 +50,20 @@ class BlockManagerMaster(
+  /** Register the BlockManager's id with the driver. The input BlockManagerId does not contain
+   * topology information. This information is obtained from the master and we respond with an
+   * updated BlockManagerId fleshed out with this information.
+   *
```
extra newline

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala:
```diff
@@ -103,8 +121,11 @@ private[spark] object BlockManagerId {
    */
-  def apply(execId: String, host: String, port: Int): BlockManagerId =
-    getCachedBlockManagerId(new BlockManagerId(execId, host, port))
+  def apply(execId: String,
```
The style here is inconsistent, can you move execId to the next line?

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -50,12 +50,20 @@ class BlockManagerMaster(
+   */
   def registerBlockManager(
-      blockManagerId: BlockManagerId, maxMemSize: Long, slaveEndpoint: RpcEndpointRef): Unit = {
+                            blockManagerId: BlockManagerId,
```
There is also a style issue here

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMessages.scala:
```diff
@@ -111,4 +111,6 @@ private[spark] object BlockManagerMessages {
 
   case class HasCachedBlocks(executorId: String) extends ToBlockManagerMaster
+
+  case class GetTopologyInfo(host: String) extends ToBlockManagerMaster
```
Is this message still used?

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,76 @@
+   *                randomness if needed.
+   * @return A prioritized list of peers. Lower the index of a peer, higher its priority
+   */
+  def prioritize(blockManagerId: BlockManagerId,
```
style

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,76 @@
+   *                randomness if needed.
+   * @return A prioritized list of peers. Lower the index of a peer, higher its priority
+   */
+  override def prioritize(blockManagerId: BlockManagerId,
```
style

shubhamchopra committed:
Fixing style issues.

shubhamchopra committed:
Fixing style issues.

shubhamchopra committed:
Adding documentation to clarify on topology.

shubhamchopra committed:
Adding a simple file based topology-mapper.

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -126,6 +138,10 @@ class BlockManagerMasterEndpoint(
       }
   }
 
+  private def getTopologyInfoForHost(host: String): String = {
```
inline?

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,78 @@
+/**
+ * ::DeveloperApi::
+ * BlockReplicationPrioritization provides logic for prioritizing a sequence of peers for
+ * replicating blocks
```
add: BlockManager will replicate to each peer returned in order until the desired replication order is reached. If a replication fails, prioritize() will be called again to get a fresh prioritization.

ericl commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+
+  override def getTopologyForHost(hostname: String): String = {
+    val topology = topologyMap.get(hostname)
+    if(topology.isDefined) {
```
space after the if

ericl commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+class FileBasedTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
+
+  val topologyFile = conf.getOption("spark.replication.topologyawareness.topologyfile")
+  require(topologyFile.isDefined, "Please provide topology file for FileBasedTopologyMapper.")
```
newline

ericl commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+/**
+ * A simple file based topology mapper. This expects topology information provided as a
+ * [[java.util.Properties]] file. The name of the file is obtained from SparkConf property
+ * `spark.replication.topologyawareness.topologyfile`. To use this topology mapper, set the
```
s/topologyfile/topologyFile

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,88 @@ private[spark] class BlockManager(
+        logTrace(s"Replicated $blockId of ${data.size} bytes to $peer" +
+          s" in ${System.currentTimeMillis - onePeerStartTime} ms")
+        // the block was replicated, lets update state and move ahead
+
```
extra newline

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,88 @@ private[spark] class BlockManager(
+          classTag)
+        logTrace(s"Replicated $blockId of ${data.size} bytes to $peer" +
+          s" in ${System.currentTimeMillis - onePeerStartTime} ms")
+        // the block was replicated, lets update state and move ahead
```
don't need this comment

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,88 @@ private[spark] class BlockManager(
+    var peersFailedToReplicateTo = Set.empty[BlockManagerId]
+    var numFailures = 0
+
+    while(numFailures <= maxReplicationFailures
```
The indentation still looks funny here. Should the && be on the trailing part of the line?

ericl commented:
LGTM. Just style comments

shubhamchopra committed:
Fixing style issues.

ericl commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -138,9 +138,7 @@ class BlockManagerMasterEndpoint(
-  private def getTopologyInfoForHost(host: String): String = {
-    topologyMapper.getTopologyForHost(host)
-  }
+  private def getTopologyInfoForHost(h: String): String = topologyMapper.getTopologyForHost(h)
```
I meant inline the function into the caller. It only seems to be used in one place.

ericl commented:
still lgtm

shubhamchopra committed:
Inlining call to get topology.

shubhamchopra committed:
converting to not a and not b

ericl commented:
@rxin

SparkQA commented:
**[Test build #3225 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3225/consoleFull)** for PR 13152 at commit [`9b8ce32`](https://github.com/apache/spark/commit/9b8ce3229d0cff64e77d55563cec3cc3cda29182).

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,81 @@
+   *         to make this topology information nested.
+   *         For example : ‘/myrack/myhost’, where ‘/’ is the topology delimiter,
+   *         ‘myrack’ is the topology identifier, and ‘myhost’ is the individual host.
+   *         This function only returns the topology information without the hostname.
```
can you document what an empty string means?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,80 @@
+   * @return A prioritized list of peers. Lower the index of a peer, higher its priority
+   */
+  override def prioritize(
+    blockManagerId: BlockManagerId,
```
so the Spark style for indentation is to have 4 spaces for function arguments, i.e.
```scala
override def prioritize(
    blockManagerId: BlockManagerId,,
    peers: Seq[BlockManagerId],
    peersReplicatedTo: Set[BlockManagerId],
    blockId: BlockId): Seq[BlockManagerId] = {
  val random = new Random(blockId.hashCode)
  ...
}
```

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,80 @@
+}
+
+@DeveloperApi
+class DefaultBlockReplicationPrioritization
```
instead of Default, I'd call this RandomBlockReplicationPrioritization to better reflect what it does.

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,80 @@
+   */
+  def prioritize(
+    blockManagerId: BlockManagerId,
+    peers: Seq[BlockManagerId],
```
is passing in all the peers a performance concern?

rxin replied:
also rather than a full prioritization, can we also pass in a number of replicas wanted and just return a list there?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -160,8 +163,25 @@ private[spark] class BlockManager(
+        "org.apache.spark.storage.DefaultBlockReplicationPrioritization")
+      val clazz = Utils.classForName(priorityClass)
+      val ret = clazz.newInstance.asInstanceOf[BlockReplicationPrioritization]
+      logInfo(s"Using $priorityClass for prioritizing peers")
```
```
Using $priorityClass for block replication policy
```

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMaster.scala:
```diff
@@ -50,12 +50,20 @@ class BlockManagerMaster(
+   */
   def registerBlockManager(
-      blockManagerId: BlockManagerId, maxMemSize: Long, slaveEndpoint: RpcEndpointRef): Unit = {
+    blockManagerId: BlockManagerId,
```
indent 4 spaces

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -298,7 +310,17 @@ class BlockManagerMasterEndpoint(
 
-  private def register(id: BlockManagerId, maxMemSize: Long, slaveEndpoint: RpcEndpointRef) {
+  private def register(dummyId: BlockManagerId,
+    maxMemSize: Long,
```
4 space indent, and put all the arguments on its own line, e.g.
```
private def register(
    dummyId: BlockManagerId,
    maxMemSize: Long,
    slaveEndpoint: RpcEndpointRef): BlockManagerId = {
  ...
}
```

rxin replied:
also instead of dummyId, I'd call it "idWithoutTopologyInfo"

rxin replied:
Can you also add a method doc saying this returns the same id with topology information attached?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala:
```diff
@@ -101,10 +117,18 @@ private[spark] object BlockManagerId {
-  def apply(execId: String, host: String, port: Int): BlockManagerId =
-    getCachedBlockManagerId(new BlockManagerId(execId, host, port))
+  def apply(
+    execId: String,
```
4 space indent here too

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala:
```diff
@@ -69,24 +72,37 @@ class BlockManagerId private (
     host_ = in.readUTF()
     port_ = in.readInt()
+    val isTopologyInfoAvailable = in.readBoolean()
+    topologyInfo_ = if (isTopologyInfoAvailable) {
```
it might be more clear to do
```
if (isTopologyInfoAvailable) {
  topologyInfo_ = Option(in.readUTF())
}  else {
  topologyInfo_ = None
}
```

or

```
topologyInfo_ = if (isTopologyInfoAvailable) Option(in.readUTF()) else None
```

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
-      data: ChunkedByteBuffer,
-      level: StorageLevel,
-      classTag: ClassTag[_]): Unit = {
+    blockId: BlockId,
```
reset the change here - use 4 space indent

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+   * Replicate block to another node. Note that this is a blocking call that returns after
    * the block has been replicated.
+   *
+   * @param blockId
```
remove these params unless you really are going to document them.

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPrioritization.scala:
```diff
@@ -0,0 +1,80 @@
+ * again to get a fresh prioritization.
+ */
+@DeveloperApi
+trait BlockReplicationPrioritization {
```
can we just name this BlockReplicationPolicy?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.storage
 import java.io._
 import java.nio.ByteBuffer
 
+import scala.annotation.tailrec
```
is this used anywhere?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+    var numFailures = 0
+
+    while(numFailures <= maxReplicationFailures &&
+      !peersForReplication.isEmpty &&
```
might be good to indent 2 more spaces here to so it doesn't look like the body of the loop

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+      peersReplicatedTo.size != numPeersToReplicateTo) {
+      val peer = peersForReplication.head
+      try {
+        val onePeerStartTime = System.currentTimeMillis
```
i understand you didn't introduce this problem, but do you mind changing this to use nanoTime while you are at it? 

For more information see https://github.com/databricks/scala-style-guide#misc_currentTimeMillis_vs_nanoTime

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+        peersForReplication = peersForReplication.tail
+        peersReplicatedTo += peer
+      } catch {
+        case e: Exception =>
```
catch NonFatal here rather than general Exception?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+        peersReplicatedTo += peer
+      } catch {
+        case e: Exception =>
+          logWarning(s"Failed to replicate $blockId to $peer, failure #$numFailures", e)
```
this should be error level log, shouldn't it?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+    var peersForReplication =
+      blockReplicationPrioritizer.prioritize(blockManagerId, getPeers(false), Set.empty, blockId)
+    var peersReplicatedTo = Set.empty[BlockManagerId]
+    var peersFailedToReplicateTo = Set.empty[BlockManagerId]
```
can you make it explicit to use a mutable HashSet here?

rxin replied:
for both peersReplicatedTo and peersFailedToReplicateTo

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1088,109 +1108,86 @@ private[spark] class BlockManager(
+    while(numFailures <= maxReplicationFailures &&
+      !peersForReplication.isEmpty &&
+      peersReplicatedTo.size != numPeersToReplicateTo) {
+      val peer = peersForReplication.head
```
one danger with the head/tail operation here is that it is possible for this to have very bad performance if the underlying data structures returned by prioritize is expensive to do these kind of operations. Should we make it return either a List (then you are fine with head/tail), or an Array (then you can just use indexes)?

SparkQA commented:
**[Test build #3225 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3225/consoleFull)** for PR 13152 at commit [`9b8ce32`](https://github.com/apache/spark/commit/9b8ce3229d0cff64e77d55563cec3cc3cda29182).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

shubhamchopra committed:
1. Style modifications as suggested by @rxin. 2. Changing the BlockReplicationPolicy api to take the number of peers needed and adding a sampling algo linear in time and space along with test cases.

SparkQA commented:
**[Test build #3232 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3232/consoleFull)** for PR 13152 at commit [`a0ed1e6`](https://github.com/apache/spark/commit/a0ed1e648a343f123bc41c1bdb94766b5a1e7938).

SparkQA commented:
**[Test build #3232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3232/consoleFull)** for PR 13152 at commit [`a0ed1e6`](https://github.com/apache/spark/commit/a0ed1e648a343f123bc41c1bdb94766b5a1e7938).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPolicy.scala:
```diff
@@ -0,0 +1,112 @@
+      peers: Seq[BlockManagerId],
+      peersReplicatedTo: mutable.HashSet[BlockManagerId],
+      blockId: BlockId,
+      numPeersToReplicateTo: Int): List[BlockManagerId]
```
numReplicas?

rxin commented on core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:
```diff
@@ -347,6 +347,47 @@ class BlockManagerReplicationSuite extends SparkFunSuite
+    val blockId = "test-block"
+
+    (1 to 10).foreach {numReplicas =>
+      logInfo(s"Num replicas : $numReplicas")
```
are these logging information useful in tests?

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+ */
+@DeveloperApi
+class FileBasedTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
+  val topologyFile = conf.getOption("spark.replication.topologyawareness.topologyfile")
```
I was looking at existing configs - can we change the config to
```
spark.storage.replication.topologyFile
```

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+@DeveloperApi
+class FileBasedTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
+  val topologyFile = conf.getOption("spark.replication.topologyawareness.topologyfile")
+  require(topologyFile.isDefined, "Please provide topology file for FileBasedTopologyMapper.")
```
"Please specify topology file via spark.storage.replication.topologyFile for FileBasedTopologyMapper."

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+class DefaultTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
+  override def getTopologyForHost(hostname: String): Option[String] = {
+    logDebug(s"Got a request for $hostname")
+    Some("DefaultRack")
```
should this return None instead?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -160,8 +164,25 @@ private[spark] class BlockManager(
+    blockReplicationPolicy = {
+      val priorityClass = conf.get(
+        "spark.replication.topologyawareness.prioritizer",
+        "org.apache.spark.storage.RandomBlockReplicationPolicy")
```
classOf[RandomBlockReplicationPolicy].getName

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+}
+
+@DeveloperApi
+class DefaultTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
```
add a simple classdoc here - something like "A TopologyMapper that assumes all nodes are in the same rack"

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala:
```diff
@@ -55,10 +55,22 @@ class BlockManagerMasterEndpoint(
 
+  private val topologyMapper = {
+    val topologyMapperClassName = conf.get(
+      "spark.replication.topologyawareness.topologyMapper",
```
```
spark.storage.replication.topologyMapper
```

rxin commented on core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:
```diff
@@ -37,6 +35,8 @@ import org.apache.spark.serializer.{KryoSerializer, SerializerManager}
 import org.apache.spark.shuffle.sort.SortShuffleManager
 import org.apache.spark.storage.StorageLevel._
 
+import scala.collection.mutable
```
i think this will fail the style checker for import order

rxin commented on core/src/main/scala/org/apache/spark/storage/TopologyMapper.scala:
```diff
@@ -0,0 +1,82 @@
+ * @param conf SparkConf object
+ */
+@DeveloperApi
+class FileBasedTopologyMapper(conf: SparkConf) extends TopologyMapper(conf) with Logging {
```
can we add a unit test for FileBasedTopologyMapper itself?

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPolicy.scala:
```diff
@@ -0,0 +1,112 @@
+ * again to get a fresh prioritization.
+ */
+@DeveloperApi
+trait BlockReplicationPolicy {
```
can we create a BlockReplicationPolicySuite and add unit tests for RandomBlockReplicationPolicy? You'd want to verify it is returning you the right number of peers and make sure they don't duplicate each other ...

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockReplicationPolicy.scala:
```diff
@@ -0,0 +1,112 @@
+   * @return list of m random unique indices
+   */
+  private def getSampleIds(n: Int, m: Int, r: Random): List[Int] = {
+    val indices = (n - m + 1 to n).foldLeft(Set.empty[Int]) {case (s, i) =>
```
this is a bit difficult to read (actually very rarely do foldLeft make code easy to read in my experience).

Perhaps just write it in a more imperative way?

shubhamchopra replied:
I changed the variable names a little, to make it more clear that the left argument is the set that we are accumulating on.
This is just a 2 line usage, I am fine either ways. Using folds, though, is a more general and the recommended idiomatic way of writing accumulating loops of this kind in Scala. In my experience, I have found imperative code to be a lot more error prone and harder to reason about. I generally stick with idiomatic Scala, unless practicality dictates otherwise.

rxin commented:
Thanks - this looks pretty good!

I've triggered a new Jenkins run and also left some small comments. It would be great to add some unit tests (not integration tests) for two of the classes introduced.

shubhamchopra committed:
Incorporating corrections and suggestions by @rxin

shubhamchopra commented:
Thanks for the suggestions. I have corrected the style check errors and verified that locally, so hopefully there are not more style errors. I have also done a couple of modifications per your suggestions:
1. Added TopologyMapperSuite and BlockReplicationPolicySuite for unit tests on these classes.
2. Changed config parameters names as suggested.

shubhamchopra commented:
Rebased to master to resolve merge conflicts

SparkQA commented:
**[Test build #3291 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3291/consoleFull)** for PR 13152 at commit [`632d043`](https://github.com/apache/spark/commit/632d0436ca701031e03bd141b95d4b0bb5544150).

rxin commented:
LGTM - sorry that this has taken a while. I will merge once tests pass.

Also cc @zsxwing for his attention.

SparkQA commented:
**[Test build #3291 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/3291/consoleFull)** for PR 13152 at commit [`632d043`](https://github.com/apache/spark/commit/632d0436ca701031e03bd141b95d4b0bb5544150).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

rxin commented:
Merging in master.
